### ISSUE TITLE
Lint for ambiguous method defined in trait impl for struct & on struct itself

### ISSUE THREAD

**Issue by 0xngold** [rust-lang/rust-clippy#11499]
### What it does

This was inspired by [rust-lang #37741](https://github.com/rust-lang/rust/issues/37741). In short, it's very easy to accidentally override a trait method impl. by adding a method of the same name to a struct impl.. This causes unqualified call sites to land in the struct's method instead of the trait method impl for the struct. (Generally call sites will be unqualified because naming conflicts between traits are rare, so this will happen silently.) Since clippy generally warns for potentially incorrect code, this feels like a good candidate for a new lint.

The problematic situation is shown in [this playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=24851dd74560c3418e3f7e416a3c05b0). For simplicity, here's the code directly:

```rust
trait DataExt {
    fn ambiguous(&self);
}

struct BaseData {}

impl BaseData {
    // This method shadows DataExt::ambiguous(self).
    fn ambiguous(&self) {
        println!("struct impl");
    }
}

impl DataExt for BaseData {
    fn ambiguous(&self) {
        println!("trait impl");
    }
}

fn main() {
    let data = BaseData{};
    
    // This calls BaseData::ambiguous instead of DataExt::ambiguous & generates no warnings,
    // despite the trait being in scope.
    data.ambiguous();
}
```

The proposal would be to flag `BaseData::ambiguous` as creating ambiguity with `DataExt::ambiguous`. In other words, if a struct `S` implements a trait `T` with method `T::foo`, then a method `S::foo` should generate an ambiguity warning.

### Advantage

- warns about unintentional "overriding" of a trait method already defined on a struct.
- encourages writing easier to understand code

### Drawbacks

- complex object hierarchies may intentionally want this behavior (this is why we probably want the lint to be default allowed)

### Example

The lint should trigger as follows:

```rust
trait DataExt {
    fn ambiguous(&self);
}

struct BaseData {}

impl BaseData {
    // This method shadows DataExt::ambiguous(self) and should fail the lint.
    fn ambiguous(&self) {
        println!("struct impl");
    }
}

impl DataExt for BaseData {
    fn ambiguous(&self) {
        println!("trait impl");
    }
}
```

The developer could then rewrite as:

```rust
trait DataExt {
    fn ambiguous(&self);
}

struct BaseData {}

impl BaseData {
    fn no_longer_ambiguous(&self) {
        println!("struct impl");
    }
}

impl DataExt for BaseData {
    fn ambiguous(&self) {
        println!("trait impl");
    }
}
```

**Comment by sjwang05** [comment]
@rustbot claim

I'm not sure how difficult this would be since it's my first time contributing, but I'd be happy to work on a lint for this

**Comment by Centri3** [comment]
Hey @sjwang05 !

If you need help, make sure to ask here or on zulip :)

**Comment by sjwang05** [comment]
Got it, thanks 👍

**Comment by sjwang05** [comment]
Draft PR is open here: #11528

**Comment by Jarcho** [comment]
This already exists as `same_name_method`.
